keyv/Ceph.cpp

```diff
@@ -32,262 +35,215 @@ lunchbox::PluginRegisterer<Ceph> registerer;
+    if (ret < 0)
+    {
+        std::cerr << "Get failed: " << ::strerror(-ret) << std::endl;
+        return std::string();
+    }
 
-        asyncRead.op.reset(librados::Rados::aio_create_completion());
-        uint64_t size = sizeHint;
-        if (size == 0)
-        {
-            time_t time;
-            const int stat = _context.stat(key, &size, &time);
-            if (stat < 0 || size == 0)
-            {
-                std::cerr << "Stat " << key << " failed: " << ::strerror(-stat)
-                          << std::endl;
-                return false;
-            }
-        }
+    auto pos = map.find(key);
+    if (pos == map.end())
+    {
+        return std::string();
+    }
 
-        const int read =
-            _context.aio_read(key, asyncRead.op.get(), &asyncRead.bl, size, 0);
-        if (read >= 0)
-            return true;
+    char* data = pos->second.c_str();
+    std::string str(data, pos->second.length());
+    return str;
+}
 
-        std::cerr << "Fetch failed: " << ::strerror(-read) << std::endl;
-        return false;
+inline void Ceph::takeValues(const lunchbox::Strings& keys,
+                             const ValueFunc& func) const
+{
+    IOMap map;
+    int ret = _context.omap_get_vals_by_keys(
+        _storeName, std::set<std::string>(keys.begin(), keys.end()), &map);
+    if (ret < 0)
+    {
+        std::cerr << "Take failed: " << ::strerror(-ret) << std::endl;
+        return;
     }
 
-    bool _flush(const size_t maxPending)
+    for (auto& pair : map)
     {
-        if (maxPending == 0)
-        {
-            const int flushAll = _context.aio_flush();
-            while (!_writes.empty())
-            {
-                delete _writes.front();
-                _writes.pop_front();
-            }
-            if (flushAll >= 0)
-                return true;
-
-            std::cerr << "Flush all writes failed: " << ::strerror(-flushAll)
-                      << std::endl;
-            return false;
-        }
+        librados::bufferlist bl = pair.second;
+        if (bl.length() == 0)
+            continue;
+
+        char* data = pair.second.c_str();
+        char* copy = (char*)malloc(bl.length());
+        std::copy(data, data + bl.length(), copy);
+        func(pair.first, copy, bl.length());
+    }
+}
 
-        bool ok = true;
-        while (_writes.size() > maxPending)
-        {
-            _writes.front()->wait_for_complete();
-            const int write = _writes.front()->get_return_value();
-            if (write < 0)
-            {
-                std::cerr << "Finish write failed: " << ::strerror(-write)
-                          << std::endl;
-                ok = false;
-            }
-            delete _writes.front();
-            _writes.pop_front();
-        }
-        return ok;
+inline void Ceph::getValues(const lunchbox::Strings& keys,
+                            const ConstValueFunc& func) const
```

Comment:
Is it possible to refactor this function and takeValues to extract the common code?